lib/iris/_lazy_data.py

```diff
@@ -45,102 +37,18 @@ def is_lazy_data(data):
-            # Grab a fill value, in case this is just a converted masked array.
-            fill_value = getattr(data, 'fill_value', None)
-            # Realise dask array.
-            data = data.compute()
-            # Convert NaN arrays into masked arrays for Iris' consumption.
-            mask = np.isnan(data)
-            if np.all(~mask):
-                mask = None
-            data = np.ma.masked_array(data, mask=mask,
-                                      fill_value=fill_value)
-    return data
-
-
-# A magic value, borrowed from biggus
-_MAX_CHUNK_SIZE = 8 * 1024 * 1024 * 2
-
-
-def as_lazy_data(data):
-    """
-    Return a lazy equivalent of the argument, as a lazy array.
-
-    For an existing lazy array, return it unchanged.
-    Otherwise, return the argument wrapped with dask.array.from_array.
-    This assumes the underlying object has numpy-array-like properties.
-
-    .. Note::
-
-        For now at least, chunksize is set to an arbitrary fixed value.
-
-    """
-    if not is_lazy_data(data):
-        # record the original fill value.
-        fill_value = getattr(data, 'fill_value', None)
-        if isinstance(data, np.ma.MaskedArray):
-            # Use with NaNs replacing the mask.
-            data = array_masked_to_nans(data)
-        data = da.from_array(data, chunks=_MAX_CHUNK_SIZE)
-        # Attach any fill value to the dask object.
-        # Note: this is not passed on to dask arrays derived from this one.
-        data.fill_value = fill_value
-    elif not hasattr(data, 'fill_value'):
-        data.fill_value = None  # make it look more like a biggus Array ?
-    return data
-
-
-def array_masked_to_nans(array):
+def array_masked_to_nans(array, mask=None):
     """
     Convert a masked array to a normal array with NaNs at masked points.
-
     This is used for dask integration, as dask does not support masked arrays.
     Note that any fill value will be lost.
-
-    """
-    if np.ma.is_masked(array):
-        # Array has some masked points : use unmasked near-equivalent.
-        if array.dtype.kind == 'f':
-            # Floating : convert the masked points to NaNs.
-            array = array.filled(np.nan)
-        else:
-            # Integer : no conversion (i.e. do *NOT* fill with fill value)
-            # array = array.filled()
-            array = array.data
-    else:
-        # Ensure result is not masked (converts arrays with empty masks).
-        if isinstance(array, np.ma.MaskedArray):
-            array = array.data
-    return array
-
-
-def array_nans_to_masked(array):
-    """
-    Convert an array into a masked array, masking any NaN points.
-
     """
-    if (not isinstance(array, np.ma.masked_array) and
-            array.dtype.kind == 'f'):
-        mask = np.isnan(array)
-        if np.any(mask):
-            # Turn any unmasked array with NaNs into a masked array.
-            array = np.ma.masked_array(array, mask=mask)
+    if mask is None:
+        mask = array.mask
```

Comment:
@marqh Of course if you don't pass in a masked array, then this will raise an exception ...

... every time this function is called (pretty) much, it's always within an `if` statement to check to see if the `array` is a masked array.

Why don't we refactor this, so that any array can be passed in. If the `array` is not a `MaskedArray`  then just pass it through. If it's a `MaskedArray` then do the NaN-ifycation, as is.

That would save on the use of `if isinstance(array, np.ma.MaskedArray) .... array_masked_to_nans(...)` pattern.

If you buy into that, then I'd simply suggest calling this function `array_to_nans` ... but that's minor to my previous points.